service_configuration_lib/spark_config.py

```diff
@@ -460,19 +467,23 @@ def _get_k8s_spark_env(
         'spark.kubernetes.authenticate.clientKeyFile': f'{K8S_AUTH_FOLDER}/{paasta_cluster}-client.key',
         'spark.kubernetes.authenticate.clientCertFile': f'{K8S_AUTH_FOLDER}/{paasta_cluster}-client.crt',
         'spark.kubernetes.container.image.pullPolicy': 'Always',
-        'spark.kubernetes.executor.label.yelp.com/paasta_service': paasta_service,
-        'spark.kubernetes.executor.label.yelp.com/paasta_instance': paasta_instance,
-        'spark.kubernetes.executor.label.yelp.com/paasta_cluster': paasta_cluster,
-        'spark.kubernetes.executor.label.paasta.yelp.com/service': paasta_service,
-        'spark.kubernetes.executor.label.paasta.yelp.com/instance': paasta_instance,
-        'spark.kubernetes.executor.label.paasta.yelp.com/cluster': paasta_cluster,
+        'spark.kubernetes.executor.label.yelp.com/paasta_service': _paasta_service,
+        'spark.kubernetes.executor.label.yelp.com/paasta_instance': _paasta_instance,
+        'spark.kubernetes.executor.label.yelp.com/paasta_cluster': _paasta_cluster,
+        'spark.kubernetes.executor.label.paasta.yelp.com/service': _paasta_service,
+        'spark.kubernetes.executor.label.paasta.yelp.com/instance': _paasta_instance,
+        'spark.kubernetes.executor.label.paasta.yelp.com/cluster': _paasta_cluster,
         'spark.kubernetes.node.selector.yelp.com/pool': paasta_pool,
         'spark.kubernetes.executor.label.yelp.com/pool': paasta_pool,
         **_get_k8s_docker_volumes_conf(volumes),
     }
     return spark_env
 
 
+def _get_k8s_resource_name_prefix(resource_name: str):
+    return re.sub('-+', '-', re.sub('[^a-z0-9\\-]', '-', resource_name.strip().lower()))
```

Comment:
Can you add some examples for this?
It looks like this expression is:
* Replacing all consecutive --'s to a single - and mostly any alpha-numeric characters except \ to -?
Is adding \\ required?

I also noticed that the failing label includes "paasta-spark-etl-runner" as prefix in the beginning. Do you think, it would be useful to truncate or replace it with shorter string to start with. That way it is less likely to have conflicts with 2 services having same label name after truncation? 
We know spark-etl is always on paasta and is a runner. So just maybe, 'spark-etl' or even 'etl'.
Thoughts?
Adding a docstring here would be useful. Took me a while to understand what is happening here :) 

I think adding a warning log line would be helpful here for before and after resource-name string if it changed or truncated.


Probably not in this review, but a unit-test would be useful

Reply:
>  mostly any alpha-numeric characters except \ to -
including `\`, I'm actually just following the rules for those resource names, I haven't encountered these cases yet. I might actually add `_` which is getting replaced to `-` but it may cause confusion.

> I also noticed that the failing label includes "paasta-spark-etl-runner" as prefix in the beginning.
No it's this case, the failing label is not in the log unfortunately (that's what made this bug tricky), but it's in the values that are set for `spark.kubernetes.executor.label.yelp.com/paasta_instance` and `spark.kubernetes.executor.label.paasta.yelp.com/instance` which in Tron follow the rule of `tron_job_name.tron_job_action`, set [here](https://sourcegraph.yelpcorp.com/Yelp/paasta@91a8540d9c2e5c5fc50b1048639b50f5d231e912/-/blob/paasta_tools/cli/cmds/spark_run.py#L395-398)

> That way it is less likely to have conflicts with 2 services having same label name after truncation?
good point, this might still apply in this case. Although these labels are not set to uniquely identify a job (more like filtering down a job) with the initial substring part it should still be able to filter adequately. In the long run, I'm planning after this change to introduce new tron specific labels, like `spark.kubernetes.executor.label.tron.yelp.com/job_name` and `spark.kubernetes.executor.label.tron.yelp.com/job_action`, but that's part of another change.

> I think adding a warning log line would be helpful here for before and after resource-name string if it changed or truncated.
these labels are purely used in case you want to filter down jobs with `kubectl-pnw-devc -n paasta-spark get pods -o wide -l paasta.yelp.com/instance=instance_name`. 

>  Adding a docstring here would be useful. Took me a while to understand what is happening here :)
good point! will add!